okhttp/src/main/java/io/grpc/okhttp/DelegatingFrameWriter.java

```diff
@@ -0,0 +1,204 @@
+/*
+ * Copyright 2018 The gRPC Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.grpc.okhttp;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import com.google.common.base.MoreObjects;
+import io.grpc.okhttp.internal.framed.ErrorCode;
+import io.grpc.okhttp.internal.framed.FrameWriter;
+import io.grpc.okhttp.internal.framed.Header;
+import io.grpc.okhttp.internal.framed.Settings;
+import java.io.IOException;
+import java.net.Socket;
+import java.util.List;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import okio.Buffer;
+
+class DelegatingFrameWriter implements FrameWriter {
+
+  private static final Logger log = Logger.getLogger(OkHttpClientTransport.class.getName());
+
+  private final TransportExceptionHandler transportExceptionHandler;
+
+  private FrameWriter frameWriter;
+  private Socket socket;
+
+  DelegatingFrameWriter(
+      FrameWriter frameWriter, Socket socket, TransportExceptionHandler transportExceptionHandler) {
+    this.frameWriter = checkNotNull(frameWriter, "frameWriter");
+    // the socket is needed for closing.
+    this.socket = checkNotNull(socket, "socket");
+    this.transportExceptionHandler =
+        checkNotNull(transportExceptionHandler, "transportExceptionHandler");
+  }
+
+  @Override
+  public void connectionPreface() {
+    try {
+      frameWriter.connectionPreface();
+    } catch (Exception e) {
+      transportExceptionHandler.onException(e);
+    }
+  }
+
+  @Override
+  public void ackSettings(Settings peerSettings) {
+    try {
+      frameWriter.ackSettings(peerSettings);
+    } catch (Exception e) {
+      transportExceptionHandler.onException(e);
+    }
+  }
+
+  @Override
+  public void pushPromise(int streamId, int promisedStreamId, List<Header> requestHeaders) {
+    try {
+      frameWriter.pushPromise(streamId, promisedStreamId, requestHeaders);
+    } catch (Exception e) {
+      transportExceptionHandler.onException(e);
+    }
+  }
+
+  @Override
+  public void flush() {
+    try {
+      frameWriter.flush();
+    } catch (Exception e) {
+      transportExceptionHandler.onException(e);
+    }
+  }
+
+  @Override
+  public void synStream(boolean outFinished, boolean inFinished, int streamId,
```

Comment:
This line wrapping is kinda funky.  I prefer to put all the args on their own line if it would wrap.  This also follows the rectangle rule.